tests/language-feature/generics/irwarray.slang

```diff
@@ -0,0 +1,34 @@
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECK): -shaderobj -output-using-type
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECK): -vk -shaderobj -output-using-type
+
+void writeToArray<U, T : IRWArray<U>>(inout T array, int index, U value) { array[index] = value; }
+void writeToBuffer<U, T : IRWArray<U>>(T array, int index, U value) { array[index] = value; }
+U readFromArray<U, T:IArray<U>>(T array, int index) { return array[index]; }
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=outputBuffer
+RWStructuredBuffer<float> outputBuffer;
+
+[numthreads(1, 1, 1)]
+void computeMain(int3 dispatchThreadID: SV_DispatchThreadID)
+{
+    float arr[3] = { 1.0, 2.0, 3.0 };
+    float4 v = float4(1.0, 2.0, 3.0, 4.0);
+    float2x2 m = float2x2(1.0, 2.0, 3.0, 4.0);
+
+    // CHECK: 1.0
+    writeToBuffer(outputBuffer, 0, 1.0f);
+
+    // CHECK: 4.0
+    writeToArray(arr, 0, 4.0f);
+    outputBuffer[1] = readFromArray(arr, 0);
```

Comment:
I am little confused by the inconsistency with `writeToBuffer`.
If I do `writeToBuffer(outputBuffer, ...)`, it will modify `outputBuffer`.
But if I do `writeToBuffer(arr, ...)`, it wouldn't modify `arr`.

When I look at the function signature alone, it shouldn't modify the first argument because it doesn't have `out` nor `inout` modifier.
```
void writeToBuffer<U, T : IRWArray<U>>(T array, int index, U value)
```
When used with RWStructuredBuffer, `T array` works as if it is `inout T array`, which seems inconsistent.

Reply:
Yeah, this is a little unfortunate, but at the moment there isn't a better solution.

RWStructuredBuffer<T> itself should be viewed as `T*`, that is a handle to an array. Therefore, The structuredbuffer itself is a non-mutable object (represented as a uniform parameter), and therefore you can't pass it into a function that expects an `inout` parameter, because you can't assign `buffer` to something else, it is readonly.

In contract, for arrays you do want to mutate its content, so `inout` is needed there.

Reply:
The essence in this is that `RWStructuredBuffer` is `Array<T>*` and not an array, but there is no correponding interface to represent a "reference to an array" right now, and even if we do, we can't construct a reference object to a real array in order to satisfy that interface anyways.